flang/lib/Evaluate/tools.cpp

```diff
@@ -82,6 +82,8 @@ auto IsVariableHelper::operator()(const Symbol &symbol) const -> Result {
   const Symbol &ultimate{symbol.GetUltimate()};
   return !IsNamedConstant(ultimate) &&
       (ultimate.has<semantics::ObjectEntityDetails>() ||
+          (ultimate.has<semantics::EntityDetails>() &&
+              ultimate.attrs().test(semantics::Attr::TARGET)) ||
```

Comment:
I am not sure I fully comprehend this change. Is it because IsVariable may be called on a symbol that is not fully resolved as being an object or a procedure yet by name resolution, but its TARGET argument allows telling it cannot be a procedure?

Is it always the case that the TARGET attribute will be set on time (could the symbol appear on a TARGET statement only late in the specification expressions that would be processed after this IsVariable call is made?).

Reply:
It's needed to cope with the case of a host-associated entity being referenced in a specification expression in an inner or module procedure.  Name resolution must (and will) process those nested specification parts before completing the processing of the specification part of the host.  Any valid `TARGET` attribute on an `EntityDetails` symbol must denote an object, and will be converted to an `ObjectEntityDetails` symbol later.  See the original motivating test case.

(In the case of an entity with `TARGET` being referenced in a specification expression later in the same specification part, the same logic applies.)

An alternative approach would be to call `ConvertToObjectEntity()` on symbols when they acquire the `TARGET` attribute, since they can't legitimately be anything but objects at that point.

Reply:
So something like the following will still be rejected with this patch, right?
```
subroutine test(x, y)
 use iso_c_binding, only : c_loc, c_ptr
  interface
    pure integer function purefunc(l)
      import
      type(c_ptr), intent(in) :: l
    end function
  end interface
 integer :: x
 integer :: y(purefunc(c_loc(x)))
 target :: x
end subroutine
```

I think nag, gfortran, ifort, and nvfortran at leat all accept it (all compiler rightfully raise an error if the target statement is removed, except nvfortran).
Maybe that is a corner case that is not worth the complexity though if it has no real useful application though.

Reply:
trying it now...

Reply:
Thanks for checking!